File: csharp/CsFind/CsFindTests/FindOptionsTests.cs
using System.Collections.Generic;
using System.Linq;
using CsFindLib;
using NUnit.Framework;
using FindOptions = CsFindLib.FindOptions;

namespace CsFindTests;

[TestFixture]
public class FindOptionsTests
{
	private readonly FindOptions _findOptions = new FindOptions();

	[Test]
	public void SettingsFromArgs_NoArgs_HasDefaultValues()
	{
		var settings = _findOptions.SettingsFromArgs(new List<string>());
		Assert.That(settings.ArchivesOnly, Is.False);
		Assert.That(settings.Debug, Is.False);
		Assert.That(settings.IncludeArchives, Is.False);
		Assert.That(settings.IncludeHidden, Is.False);
		Assert.That(settings.PrintDirs, Is.False);
		Assert.That(settings.PrintFiles);
		Assert.That(settings.PrintUsage, Is.False);
		Assert.That(settings.PrintVersion, Is.False);
		Assert.That(settings.Recursive);
		Assert.That(settings.Paths.Count, Is.EqualTo(0));
		Assert.That(settings.Verbose, Is.False);
	}

	[Test]
	public void SettingsFromArgs_ValidArgs_HasArgValues()
	{
		var args = new List<string> { "-x", "cs", "." };
		var settings = _findOptions.SettingsFromArgs(args);
		Assert.That(settings.Paths.Count, Is.EqualTo(1));
		Assert.That(settings.Paths.Contains("."));
		Assert.That(settings.InExtensions.Count, Is.EqualTo(1));
		Assert.That(settings.InExtensions.Contains(".cs"));
	}

	[Test]
	public void SettingsFromArgs_InValidArgs_ThrowsFindException()
	{
		var args = new List<string> { "-x", "cs", ".", "-Q" };
		var ex = Assert.Throws<FindException>(() => _findOptions.SettingsFromArgs(args));
		Assert.That(ex!.Message, Is.EqualTo("Invalid option: Q"));
	}

	[Test]
	public void SettingsFromJson_EqualsExpected()
	{
		var json = @"{
  ""path"": ""~/src/xfind/"",
  ""in-ext"": [""js"", ""ts""],
  ""out-dirpattern"": ""node_module"",
  ""out-filepattern"": [""temp""],
  ""debug"": true,
  ""includehidden"": true
}";
		var settings = new FindSettings();
		FindOptions.SettingsFromJson(json, settings);

		Assert.That(settings.Paths.Count, Is.EqualTo(1));
		Assert.That(settings.Paths.Contains("~/src/xfind/"));

		Assert.That(settings.InExtensions.Count, Is.EqualTo(2));
		Assert.That(settings.InExtensions.Contains(".js"));
		Assert.That(settings.InExtensions.Contains(".ts"));

		Assert.That(settings.OutDirPatterns.Count, Is.EqualTo(1));
		Assert.That(settings.OutDirPatterns.First().ToString(), Is.EqualTo("node_module"));

		Assert.That(settings.OutFilePatterns.Count, Is.EqualTo(1));
		Assert.That(settings.OutFilePatterns.First().ToString(), Is.EqualTo("temp"));

		Assert.That(settings.Debug);
		Assert.That(settings.IncludeHidden);
	}
}

File: fsharp/FsFind/FsFindTests/FindOptionsTests.fs
namespace FsFindTests

open NUnit.Framework
open FsFindLib

[<TestFixture>]
type FindOptionsTests () =

    [<SetUp>]
    member this.Setup () =
        ()

    [<Test>]
    member this.SettingsFromArgs_NoArgs_HasDefaultValues () =
        let args : string[] = [||]
        match FindOptions.SettingsFromArgs(args) with
        | Ok settings ->
            Assert.That(settings.ArchivesOnly, Is.False)
            Assert.That(settings.Debug, Is.False)
            Assert.That(settings.FollowSymlinks, Is.False)
            Assert.That(settings.IncludeArchives, Is.False)
            Assert.That(settings.IncludeHidden, Is.False)
            Assert.That(settings.Paths.Length, Is.EqualTo(0))
            Assert.That(settings.PrintDirs, Is.False)
            Assert.That(settings.PrintFiles)
            Assert.That(settings.PrintUsage, Is.False)
            Assert.That(settings.PrintVersion, Is.False)
            Assert.That(settings.Recursive)
            Assert.That(settings.SortBy, Is.EqualTo(SortBy.FilePath))
            Assert.That(settings.SortCaseInsensitive, Is.False)
            Assert.That(settings.SortDescending, Is.False)
            Assert.That(settings.Verbose, Is.False)
        | Error _ ->
            Assert.That(true, Is.False)
        ()

    [<Test>]
    member this.SettingsFromArgs_ValidArgs_HasArgValues () =
        let args = [| "-x"; "cs"; "." |]
        match FindOptions.SettingsFromArgs(args) with
        | Ok settings ->
            Assert.That(settings.InExtensions.Length, Is.EqualTo(1))
            Assert.That(settings.InExtensions |> List.exists (fun e -> e = ".cs"))
            Assert.That(settings.Paths.Length, Is.EqualTo(1))
            Assert.That(settings.Paths.Head.ToString(), Is.EqualTo("."))
        | Error _ ->
            Assert.That(true, Is.False)
        ()

    [<Test>]
    member this.SettingsFromArgs_InValidArgs_ThrowsFindException () =
        let args = [| "-x"; "cs"; "."; "-Q" |]
        match FindOptions.SettingsFromArgs(args) with
        | Ok _ ->
            Assert.That(true, Is.False)
        | Error e ->
            Assert.That(e, Is.EqualTo("Invalid option: Q"))
        ()

    [<Test>]
    member this.SettingsFromJson_EqualsExpected () =
        let json = String.concat "" [
            "{";
            "\"path\": \"~/src/xfind/\",";
            "\"in-ext\": [\"js\", \"ts\"],"
            "\"out-dirpattern\": \"node_module\",";
            "\"out-filepattern\": [\"temp\"],"
            "\"debug\": true,";
            "\"followsymlinks\": true,";
            "\"includehidden\": true"
            "}"
        ]
        match FindOptions.SettingsFromJson(json) with
        | Ok settings ->
            Assert.That(settings.Paths.Length, Is.EqualTo(1))
            Assert.That(settings.Paths.Head.ToString(), Is.EqualTo("~/src/xfind/"))
            Assert.That(settings.InExtensions.Length, Is.EqualTo(2))
            Assert.That(settings.InExtensions |> List.exists (fun e -> e = ".js"))
            Assert.That(settings.InExtensions |> List.exists (fun e -> e = ".ts"))
            Assert.That(settings.OutDirPatterns.Length, Is.EqualTo(1))
            Assert.That(settings.OutDirPatterns |> List.exists (fun p -> p.ToString() = "node_module"))
            Assert.That(settings.OutFilePatterns.Length, Is.EqualTo(1))
            Assert.That(settings.OutFilePatterns |> List.exists (fun p -> p.ToString() = "temp"))
            Assert.That(settings.Debug, Is.True)
            Assert.That(settings.FollowSymlinks, Is.True)
            Assert.That(settings.IncludeHidden, Is.True)
        | Error e ->
            Assert.That(true, Is.False)
        ()
